src/meta/processors/admin/AdminClient.cpp

```diff
@@ -5,90 +5,334 @@
  */
 
 #include "meta/processors/admin/AdminClient.h"
+#include "meta/MetaServiceUtils.h"
+#include "meta/processors/Common.h"
+
+DEFINE_int32(max_retry_times_admin_op, 3, "max retry times for admin request!");
 
 namespace nebula {
 namespace meta {
 
 folly::Future<Status> AdminClient::transLeader(GraphSpaceID spaceId,
-                                                      PartitionID partId,
-                                                      const HostAddr& leader,
-                                                      const HostAddr& dst) {
-    UNUSED(spaceId);
-    UNUSED(partId);
-    UNUSED(leader);
-    UNUSED(dst);
+                                               PartitionID partId,
+                                               const HostAddr& leader,
+                                               const HostAddr& dst) {
     if (injector_) {
         return injector_->transLeader();
     }
-    return Status::OK();
+    storage::cpp2::TransLeaderReq req;
+    req.set_space_id(spaceId);
+    req.set_part_id(partId);
+    req.set_new_leader(to(dst));
+    return getResponse(leader, std::move(req), [] (auto client, auto request) {
+               return client->future_transLeader(request);
+           }, [this] (auto&& resp) -> Status {
+               switch (resp.get_code()) {
+                   case storage::cpp2::ErrorCode::SUCCEEDED:
+                   case storage::cpp2::ErrorCode::E_LEADER_CHANGED: {
+                       return Status::OK();
+                   }
+                   default:
+                       return Status::Error("Unknown code %d",
+                                            static_cast<int32_t>(resp.get_code()));
+               }
+           });
 }
 
 folly::Future<Status> AdminClient::addPart(GraphSpaceID spaceId,
-                                                  PartitionID partId,
-                                                  const HostAddr& host,
-                                                  bool asLearner) {
-    UNUSED(spaceId);
-    UNUSED(partId);
-    UNUSED(host);
-    UNUSED(asLearner);
+                                           PartitionID partId,
+                                           const HostAddr& host,
+                                           bool asLearner) {
     if (injector_) {
         return injector_->addPart();
     }
-    return Status::OK();
+    storage::cpp2::AddPartReq req;
+    req.set_space_id(spaceId);
+    req.set_part_id(partId);
+    req.set_as_learner(asLearner);
+    return getResponse(host, std::move(req), [] (auto client, auto request) {
```

Comment:
I can not understand why need move(req) here, the req’s memory is in caller stack, and the parameter req of getResponse passed by value?

Reply:
Maybe req will be changed in the future.  So use move here.